cpp/tensorrt_llm/kernels/fusedQKNormRopeKernel.cu

```diff
@@ -0,0 +1,206 @@
+    float const base = 10000.0f // Base for RoPE computation
+)
+{
+    int const warpsPerBlock = blockDim.x / 32;
+    int const warpId = threadIdx.x / 32;
+    int const laneId = threadIdx.x % 32;
+
+    // Calculate global warp index to determine which head/token this warp processes
+    int const globalWarpIdx = blockIdx.x * warpsPerBlock + warpId;
+
+    // Total number of attention heads (Q and K)
+    int const total_qk_heads = num_heads_q + num_heads_k;
+
+    // Determine which token and head type (Q or K) this warp processes
+    int const tokenIdx = globalWarpIdx / total_qk_heads;
+    int const localHeadIdx = globalWarpIdx % total_qk_heads;
+
+    // Skip if this warp is assigned beyond the number of tokens
+    if (tokenIdx >= num_tokens)
+        return;
+
+    bool const isQ = localHeadIdx < num_heads_q;
+    int const headIdx = isQ ? localHeadIdx : localHeadIdx - num_heads_q;
+
+    int const num_heads = num_heads_q + num_heads_k + num_heads_v;
+    static_assert(head_dim % 32 == 0, "head_dim must be divisible by 32 (each warp processes one head)");
+    constexpr int numElemsPerThread = head_dim / 32;
+    constexpr int numVecPerThread = numElemsPerThread / 4;
+    float elements[numElemsPerThread];
+
+    int offsetWarp; // Offset for the warp
+    if (isQ)
+    {
+        // Q segment: token offset + head offset within Q segment
+        offsetWarp = tokenIdx * num_heads * head_dim + headIdx * head_dim;
+    }
+    else
+    {
+        // K segment: token offset + entire Q segment + head offset within K segment
+        offsetWarp = tokenIdx * num_heads * head_dim + num_heads_q * head_dim + headIdx * head_dim;
+    }
+    int offsetThread = offsetWarp + laneId * numElemsPerThread;
+
+    // Sum of squares for RMSNorm
+    float sumOfSquares = 0.0f;
+
+    // Load elements and compute sum of squares
+    int i = 0;
+
+    // Load vectorized elements
+    for (; i < numVecPerThread * 4; i += 4)
+    {
+        uint2 data;
+        data = *reinterpret_cast<uint2 const*>(&qkv[offsetThread + i]);
+
+        // Convert bfloat16 to float
+        auto vals0 = __bfloat1622float2(reinterpret_cast<__nv_bfloat162 const&>(data.x));
+        auto vals1 = __bfloat1622float2(reinterpret_cast<__nv_bfloat162 const&>(data.y));
+
+        // Store and compute sum of squares
+        elements[i] = vals0.x;
+        elements[i + 1] = vals0.y;
+        elements[i + 2] = vals1.x;
+        elements[i + 3] = vals1.y;
+
+        sumOfSquares += elements[i] * elements[i];
+        sumOfSquares += elements[i + 1] * elements[i + 1];
+        sumOfSquares += elements[i + 2] * elements[i + 2];
+        sumOfSquares += elements[i + 3] * elements[i + 3];
+    }
+    // Load remaining elements
+    for (; i < numElemsPerThread; i++)
+    {
+        elements[i] = __bfloat162float(qkv[offsetThread + i]);
+        sumOfSquares += elements[i] * elements[i];
+    }
+
+    // Reduce sum across warp
+    for (int mask = 16; mask > 0; mask /= 2)
```

Comment:
Can use `warpReduceSum` in `cpp/tensorrt_llm/common/reduceKernelUtils.cuh` directly.